pkg/imgpkg/bundle/bundle.go

```diff
@@ -19,48 +21,116 @@ const (
 )
 
 type Bundle struct {
-	plainImg     *plainimg.PlainImage
-	imgRetriever ctlimg.ImagesMetadata
+	plainImg      *plainimg.PlainImage
+	imgRetriever  ctlimg.ImagesMetadata
+	numSubBundles int
 }
 
 func NewBundle(ref string, imagesMetadata ctlimg.ImagesMetadata) *Bundle {
-	return &Bundle{plainimg.NewPlainImage(ref, imagesMetadata), imagesMetadata}
+	return &Bundle{plainimg.NewPlainImage(ref, imagesMetadata), imagesMetadata, 0}
 }
 
 func NewBundleFromPlainImage(plainImg *plainimg.PlainImage, imagesMetadata ctlimg.ImagesMetadata) *Bundle {
-	return &Bundle{plainImg, imagesMetadata}
+	return &Bundle{plainImg, imagesMetadata, 0}
 }
 
 func (o *Bundle) DigestRef() string { return o.plainImg.DigestRef() }
 func (o *Bundle) Repo() string      { return o.plainImg.Repo() }
 func (o *Bundle) Tag() string       { return o.plainImg.Tag() }
 
-func (o *Bundle) Pull(outputPath string, ui ui.UI) error {
+func (o *Bundle) Pull(outputPath string, ui goui.UI, recursive bool) error {
+	return o.pull(recursive, outputPath, "", map[string]bool{}, ui)
+}
+
+func (o *Bundle) pull(recursive bool, baseOutputPath string, bundlePath string, imagesProcessed map[string]bool, ui goui.UI) error {
 	img, err := o.checkedImage()
 	if err != nil {
 		return err
 	}
 
-	ui.BeginLinef("Pulling bundle '%s'\n", o.DigestRef())
+	if o.rootBundle(bundlePath) {
+		ui.BeginLinef("Pulling bundle '%s'\n", o.DigestRef())
+	} else {
+		ui.BeginLinef("Pulling Nested bundle '%s'\n", o.DigestRef())
+	}
 
-	err = ctlimg.NewDirImage(outputPath, img, ui).AsDirectory()
+	err = ctlimg.NewDirImage(filepath.Join(baseOutputPath, bundlePath), img, ui).AsDirectory()
 	if err != nil {
 		return fmt.Errorf("Extracting bundle into directory: %s", err)
 	}
 
-	imagesLock, err := lockconfig.NewImagesLockFromPath(filepath.Join(outputPath, ImgpkgDir, ImagesLockFile))
+	imagesLock, err := lockconfig.NewImagesLockFromPath(filepath.Join(baseOutputPath, bundlePath, ImgpkgDir, ImagesLockFile))
 	if err != nil {
 		return err
 	}
 
-	err = NewImagesLock(imagesLock, o.imgRetriever, o.Repo()).WriteToPath(outputPath, ui)
+	for _, image := range imagesLock.Images {
```

Comment:
it seems like we are looping over the entire images list even if user did not specify recursive. i see that we also have a message `To retrieve the configuration of all the nested bundles use the --recursive flag`. this is quite expensive especially for bundles that may references 1000s of images. let's not do any recursive processing if flag is not specified and drop the message.

Reply:
Not iterating over images list (to check if the image is a bundle) affects the output of the recursive bundle command. Specifically https://github.com/vmware-tanzu/carvel-imgpkg/blob/24f13393a485b185f8fadfc65e6180b7e67f5a8d/pkg/imgpkg/bundle/bundle.go#L101

@joaopapereira Thoughts here?

Can you expand on why you want to remove the message `To retrieve the configuration of all the nested bundles use the --recursive flag`. Is it to make the recursive flag less likely for folk to use it even if they don't need to? 
Another approach would be to capture your concern into the message.

Reply:
> Not iterating over images list (to check if the image is a bundle) affects the output of the recursive bundle command

so yeah, im suggesting we drop all that output so that we dont iterate over images in non-recursive case. it's a pretty  expensive operation.

Reply:
> but is this a use case we know it will exist?

this is the primary package management in kapp-controller.

both of options you are proposing are still optimizing for the wrong case imho. which use case today requires to present that message where it would be helpful?

> We can upfront fetch all the images in a threaded way and afterward just iterate over them

1000s of images in parallel is still 1000s of images. especially for systems that will be calling to imgpkg pull continiously this is very expensive.

Reply:
Ok, as discussed, 
Will drop logging
nested bundle output && the message `To retrieve the configuration of all the nested bundles use the --recursive flag`

Parallelism will be introduced as a separate piece of work and not part of this PR.